services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/CommandRouterServiceImpl.java

```diff
@@ -207,55 +218,97 @@ private Duration getSanitizedLifespan(final Duration lifespan) {
 
     @Override
     public Future<CommandRouterResult> enableCommandRouting(final List<String> tenantIds, final Span span) {
+
+        if (!running.get()) {
+            return Future.succeededFuture(CommandRouterResult.from(HttpURLConnection.HTTP_UNAVAILABLE));
+        }
+
         Objects.requireNonNull(tenantIds);
         final boolean isProcessingRequired = tenantsToEnable.isEmpty();
-        tenantsToEnable.addAll(tenantIds);
+        tenantIds.stream()
+            .filter(s -> !reenabledTenants.contains(s))
+            .filter(s -> !tenantsInProcess.contains(s))
+            .filter(s -> tenantsToEnable.stream()
+                .allMatch(entry -> !entry.one().equals(s)))
+            .forEach(s -> tenantsToEnable.addLast(Pair.of(s, 1)));
+
         if (isProcessingRequired) {
-            final Span processingSpan = tracer.buildSpan("re-enable command routing for tenants")
-                    .addReference(References.FOLLOWS_FROM, span.context())
-                    .start();
-            processTenantQueue(new ConcurrentHashSet<>(), processingSpan);
+            LOG.debug("triggering re-enabling of command routing");
+            processTenantQueue(span.context());
         }
         return Future.succeededFuture(CommandRouterResult.from(HttpURLConnection.HTTP_NO_CONTENT));
     }
 
-    private void processTenantQueue(final Set<String> processedTenants, final Span parentSpan) {
-        final String tenantId = tenantsToEnable.poll();
-        if (tenantId == null) {
-            parentSpan.finish();
-            return;
-        }
-        context.runOnContext(go -> {
-            if (processedTenants.contains(tenantId)) {
-                parentSpan.log(Map.of(
-                        Fields.MESSAGE, "skipping tenant, already processed ...",
-                        TracingHelper.TAG_TENANT_ID.getKey(), tenantId));
+    private void processTenantQueue(final SpanContext tracingContext) {
+
+        final var attempt = tenantsToEnable.pollFirst();
+        if (attempt == null) {
+            reenabledTenants.clear();
+            LOG.debug("finished re-enabling of command routing");
```

Comment:
nit pick: It's only finished if `tenantsInProcess` is empty.

Reply:
true

Reply:
I have pushed a fix